### PR TITLE
Add more type annotations

### PR DESCRIPTION
**Pull request by mfeurer**
This PR add type annotations for the remaining files in `openml` namespace.

### CODE REVIEW

File: openml/testing.py [hunk 1]
```diff
@@ -141,7 +147,9 @@ def _delete_entity_from_tracker(self, entity_type, entity):
             if entity_type == "flow":
                 delete_index = [
                     i
-                    for i, (id_, _) in enumerate(TestBase.publish_tracker[entity_type])
+                    for i, (id_, _) in enumerate(
+                        zip(TestBase.publish_tracker[entity_type], TestBase.flow_name_tracker)
```
Comment (LennartPurucker) [inline openml/testing.py:151]: Why did you add `TestBase.flow_name_tracker` here? 
Would a refactor to `for i, id_ in enumerate(TestBase.publish_tracker[entity_type])` be clearer?

Comment (mfeurer) [reply openml/testing.py:151]: Because we the flow_name in order to delete flows in the correct order (we delete flows with longer names first, because if a flow would be part of another flow, that flow needs to be deleted first, and is guaranteed to have a longer name). Speaking that out loud, it should be sufficient to delete flows by their ID, because a subflow needs to be present before the superflow is added and will therefore have a lower ID. What do you think?

Comment (mfeurer) [reply openml/testing.py:151]: @PGijsbers @janvanrijn @joaquinvanschoren would it work that we simplify the code here a bit?

Comment (PGijsbers) [reply openml/testing.py:151]: From my understanding the ID logic is correct, and sorting them based on IDs would be OK.

Comment (PGijsbers) [reply openml/testing.py:151]: Also I still don't understand what `TestBase.flow_name_tracker` does here, as its variable is not used. The only thing it does here is to make sure that it checks at most `len(TestBase.flow_name_tracker)` elements? It doesn't affect order at all?
And if you _are_ after the length cut-off, then it would be better to be more explicit, e.g.,:
```python
n_flows_to_check = len(TestBase.flow_name_tracker)
flows_to_check = TestBase.publish_tracker[entity_type][:n_flows_to_check]
... enumerate(flows_to_check) ...
```
Second, it looks like we are simply looking up the first element with `value` and then removing it from the list. There's a method for this (`list.remove`). But we can do better if we already have a set anyway:
```python
if entity_type not in TestBase.publish_tracker:
  return

if entity not in TestBase.publish_tracker[entity_type]:
  msg = f"Error deleting {entity=} of {entity_type=}: entity not found in publish tracker."
  raise ValueError(msg)

# Remove duplicates and entity:
remaining_entities = set(TestBase.publish_tracker[entity_type]) - {entity}
TestBase.publish_tracker[entity_type] = list(remaining_entities)
```

Note that the second check is only to preserve behavior that we expect the entity to be present (as indicated by the hard `[0]` index). If it doesn't matter whether or not it is present we can drop that.

Comment (PGijsbers) [reply openml/testing.py:151]: (and even then it probably makes sense to refactor further: it doesn't look like `TestBase.publish_tracker` needs to be a list in the first place, etc. But I would prefer that to be left for a separate PR)

**Review by LennartPurucker — APPROVED** [review]
Approving, to finalize the PR before the Hackathon. I will also resolve the current conflicts next.

File: openml/testing.py (hunk 1 above)
Comment (LennartPurucker) [reply openml/testing.py:160]: I noted this down and will look into it as part of a future PR and test reworks

**Comment by LennartPurucker** [comment]
The tests are now failing (they did not previously) due to server state. 
As far as I can tell, this is not related to the PR itself, and I will merge it now to prepare for the hackathon next week (where we will hopefully fix the tests).
